File: android/src/main/kotlin/com/savez/flutter_sdk/FlutterSdkPlugin.kt
package com.savez.flutter_sdk

import androidx.annotation.NonNull
import com.rlynetworkmobilesdk.EncryptedSharedPreferencesHelper
import com.rlynetworkmobilesdk.MnemonicStorageHelper
import org.kethereum.bip39.generateMnemonic
import org.kethereum.bip39.validate
import org.kethereum.bip39.dirtyPhraseToMnemonicWords
import org.kethereum.bip39.toSeed
import org.kethereum.bip39.wordlists.WORDLIST_ENGLISH
import org.kethereum.bip39.model.MnemonicWords
import org.kethereum.bip32.toKey

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.Locale

/** FlutterSdkPlugin */
const val MNEMONIC_STORAGE_KEY = "BIP39_MNEMONIC"
class FlutterSdkPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var mnemonicHelper: MnemonicStorageHelper
  private val MNEMONIC_PREFERENCE_KEY = "BIP39_MNEMONIC"

  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel: MethodChannel
  private lateinit var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_sdk")
    channel.setMethodCallHandler(this)
    this.flutterPluginBinding = flutterPluginBinding
    mnemonicHelper = MnemonicStorageHelper(flutterPluginBinding.applicationContext)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {

    when (call.method) {
        "getPlatformVersion" -> {
          result.success("Android ${android.os.Build.VERSION.RELEASE}")
        }
        "saveMnemonic" -> {
          saveMnemonic(call, result);
        }
        "getMnemonic" -> {
          getMnemonic(result)
        }
        "deleteMnemonic" -> {
          deleteMnemonic(result)
        }
        "generateNewMnemonic" -> {
          generateNewMnemonic(result)
        }
        "getPrivateKeyFromMnemonic" -> {
          getPrivateKeyFromMnemonic(call,result)
        }
        else -> {
          result.notImplemented()
        }
    }
  }

  private fun getMnemonic(result: Result) {
    mnemonicHelper.read(MNEMONIC_STORAGE_KEY) { mnemonic: String? ->
      result.success(mnemonic)
    }
  }

  private fun generateNewMnemonic(result: Result){
    val phrase = generateMnemonic(192, WORDLIST_ENGLISH)
    result.success(phrase)
  }

  private fun saveMnemonic(call: MethodCall, result: Result) {
    val mnemonic = call.argument<String>("mnemonic") ?: ""
    val useBlockStore = call.argument<Boolean>("useBlockStore")?: false
    val forceBlockStore = call.argument<Boolean>("forceBlockStore")?: false

    if (!MnemonicWords(mnemonic).validate(WORDLIST_ENGLISH)) {
      result.error("mnemonic_verification_failure", "Mnemonic is not valid", null)
    }
    mnemonicHelper.save(MNEMONIC_STORAGE_KEY, mnemonic, useBlockStore, forceBlockStore, { ->
      result.success(true)
    }, { message: String ->
      result.error("mnemonic_save_failure", message, null)
    })
  }
  private fun deleteMnemonic(result: Result) {
    mnemonicHelper.delete(MNEMONIC_STORAGE_KEY)
    result.success(true)
  }

  private fun getPrivateKeyFromMnemonic(call: MethodCall, result: Result){
    call.argument<String?>("mnemonic")?.let { mnemonic ->
      if (!MnemonicWords(mnemonic).validate(WORDLIST_ENGLISH)) {
        result.error("mnemonic_verification_failure", "mnemonic failed to pass check",null);
      }

      val words = dirtyPhraseToMnemonicWords(mnemonic)
      val seed = words.toSeed()
      val key = seed.toKey("m/44'/60'/0'/0/0")

      val privateKey = key.keyPair.privateKey.key.toByteArray()

      var intsList = mutableListOf<Int>()
      // and 0xFF fixes twos complement integer representation and
      // ensures unsigned int values pass through since
      // we cannot directly cast bytes to unsigned int
      privateKey.forEach { intsList.add(it.toInt() and 0xFF) }
      result.success(intsList)
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/FlutterSdkPlugin.swift
import Flutter
import UIKit
import Foundation


public class FlutterSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_sdk", binaryMessenger: registrar.messenger())
    let instance = FlutterSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
   
    switch call.method {
      
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "generateNewMnemonic":
      result(RlyNetworkMobileSdk().generateMnemonic())
    case "getPrivateKeyFromMnemonic":
      if let arguments = call.arguments as? [String: Any], let data = arguments["mnemonic"] as? String {
        result(RlyNetworkMobileSdk().getPrivateKeyFromMnemonic(data))
    } else {
        // Handle the case where 'arguments' or 'data' is nil
        // You might want to return an error or a default value here.
    }
      case "getMnemonic":
        result(RlyNetworkMobileSdk().getMnemonic())
      case "deleteMnemonic":
        result(RlyNetworkMobileSdk().deleteMnemonic())
      
      case "saveMnemonic":
        if let arguments = call.arguments as? [String: Any], let mnemonicToSave = arguments["mnemonic"] as? String {
          result(RlyNetworkMobileSdk().saveMnemonic(mnemonicToSave,saveToCloud: true, rejectOnCloudSaveFailure: true))
        }

      // let arguments = call.arguments as [String: Any]
      // let data = arguments["mnemonic"] as String
      // result(RlyNetworkMobileSdk().getPrivateKeyFromMnemonic(mnemonic: data))
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
